libs/Microsoft.MixedReality.WebRTC.Native/src/media/transceiver.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+#include "pch.h"
+
+#include "interop/global_factory.h"
+#include "peer_connection.h"
+#include "transceiver.h"
+#include "utils.h"
+
+namespace Microsoft::MixedReality::WebRTC {
+
+Transceiver::Transceiver(RefPtr<GlobalFactory> global_factory,
+                         MediaKind kind,
+                         PeerConnection& owner) noexcept
+    : TrackedObject(std::move(global_factory),
+                    kind == MediaKind::kAudio ? ObjectType::kAudioTransceiver
+                                              : ObjectType::kVideoTransceiver),
+      owner_(&owner),
+      kind_(kind) {
+  RTC_CHECK(owner_);
+  //< TODO
+  // RTC_CHECK(owner.sdp_semantic == webrtc::SdpSemantics::kPlanB);
+}
+
+Transceiver::Transceiver(
+    RefPtr<GlobalFactory> global_factory,
+    MediaKind kind,
+    PeerConnection& owner,
+    rtc::scoped_refptr<webrtc::RtpTransceiverInterface> transceiver) noexcept
+    : TrackedObject(std::move(global_factory),
+                    kind == MediaKind::kAudio ? ObjectType::kAudioTransceiver
+                                              : ObjectType::kVideoTransceiver),
+      owner_(&owner),
+      kind_(kind),
+      transceiver_(std::move(transceiver)) {
+  RTC_CHECK(owner_);
+  //< TODO
+  // RTC_CHECK(owner.sdp_semantic == webrtc::SdpSemantics::kUnifiedPlan);
+}
+
+Transceiver::~Transceiver() {
+  // RTC_CHECK(!owner_);
+}
+
+std::string Transceiver::GetName() const {
+  return "TODO";
+}
+
+rtc::scoped_refptr<webrtc::RtpTransceiverInterface> Transceiver::impl() const {
+  return transceiver_;
+}
+
+webrtc::RtpTransceiverDirection Transceiver::ToRtp(Direction direction) {
+  using RtpDir = webrtc::RtpTransceiverDirection;
+  static_assert((int)Direction::kSendRecv == (int)RtpDir::kSendRecv, "");
+  static_assert((int)Direction::kSendOnly == (int)RtpDir::kSendOnly, "");
+  static_assert((int)Direction::kRecvOnly == (int)RtpDir::kRecvOnly, "");
+  static_assert((int)Direction::kInactive == (int)RtpDir::kInactive, "");
+  return (RtpDir)direction;
+}
+
+Transceiver::Direction Transceiver::FromRtp(
+    webrtc::RtpTransceiverDirection rtp_direction) {
+  using RtpDir = webrtc::RtpTransceiverDirection;
+  static_assert((int)Direction::kSendRecv == (int)RtpDir::kSendRecv, "");
+  static_assert((int)Direction::kSendOnly == (int)RtpDir::kSendOnly, "");
+  static_assert((int)Direction::kRecvOnly == (int)RtpDir::kRecvOnly, "");
+  static_assert((int)Direction::kInactive == (int)RtpDir::kInactive, "");

```

Comment:
One copy of these is enough!

Reply:
I'm not sure what you mean. We need a type to expose to the API and we can't expose the WebRTC one.

Reply:
Discussed offline: removing the duplicate `static_assert` block in the other function.

Reply:
Fixed. Also added a block for `OptDirection`.